8323699: MessageFormat.toPattern() generates non-equivalent MessageFormat pattern

archiecobbs opened the pull request:
Please consider this fix to ensure that going from `MessageFormat` to pattern string via `toPattern()` and then back via `new MessageFormat()` results in a format that is equivalent to the original.

The quoting and escaping rules for `MessageFormat` pattern strings are really tricky. I admit not completely understanding them. At a high level, they work like this: The normal way one would "nest" strings containing special characters is with straightforward recursive escaping like with the `bash` command line. For example, if you want to echo `a "quoted string" example` then you enter `echo "a \"quoted string\" example"`. With this scheme it's always the "outer" layer's job to (un)escape special characters as needed. That is, the echo command never sees the backslash characters.

In contrast, with `MessageFormat` and friends, nested subformat pattern strings are always provided "pre-escaped". So to build an "outer" string (e.g., for `ChoiceFormat`) the "inner" subformat pattern strings are more or less just concatenated, and then only the `ChoiceFormat` option separator characters (e.g., `<`, `#`, `|`, etc.) are escaped.

The "pre-escape" escaping algorithm escapes `{` characters, because `{` indicates the beginning of a format argument. However, it doesn't escape `}` characters. This is OK because the format string parser treats any "extra" closing braces (where "extra" means not matching an opening brace) as plain characters.

So far, so good... at least, until a format string containing an extra closing brace is nested inside a larger format string, where the extra closing brace, which was previously "extra", can now suddenly match an opening brace in the outer pattern containing it, thus truncating it by "stealing" the match from some subsequent closing brace.

An example is the `MessageFormat` string `"{0,choice,0.0#option A: {1}|1.0#option B: {1}'}'}"`. Note the second option format string has a trailing closing brace in plain text. If you create a `MessageFormat` with this string, you see a trailing `}` only with the second option.

However, if you then invoke `toPattern()`, the result is `"{0,choice,0.0#option A: {1}|1.0#option B: {1}}}"`. Oops, now because the "extra" closing brace is no longer quoted, it matches the opening brace at the beginning of the string, and the following closing  brace, which was the previous match, is now just plain text in the outer `MessageFormat` string.

As a result, invoking `f.format(new Object{} { 0, 5 })` will return `"option A: 5"` using the original `MessageFormat` but `"option A: 5}"` from a new one created with the string from `toPattern()`.

This patch fixes this problem by adding quotes around "extra" closing braces in the subformat pattern strings.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8324172](https://bugs.openjdk.org/browse/JDK-8324172) to be approved

### Issues
 * [JDK-8323699](https://bugs.openjdk.org/browse/JDK-8323699): MessageFormat.toPattern() generates non-equivalent MessageFormat pattern (**Bug** - P4)
 * [JDK-8324172](https://bugs.openjdk.org/browse/JDK-8324172): MessageFormat.toPattern() generates non-equivalent MessageFormat pattern (**CSR**)


### Reviewers
 * [Justin Lu](https://openjdk.org/census#jlu) (@justin-curtis-lu - Committer)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17416/head:pull/17416` \
`$ git checkout pull/17416`

Update a local copy of the PR: \
`$ git checkout pull/17416` \
`$ git pull https://git.openjdk.org/jdk.git pull/17416/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17416`

View PR using the GUI difftool: \
`$ git pr show -t 17416`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17416.diff">https://git.openjdk.org/jdk/pull/17416.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17416#issuecomment-1890985229)

justin-curtis-lu commented:
Hi Archie, thanks for the proposed fix. I am still taking a look, but I wanted to demonstrate a current issue,

(Jshell with your patch)

```
var pattIn = "Test: {0,number,foo'{'#.00}";
MessageFormat mFmt = new MessageFormat(pattIn);
var pattOut  = mFmt.toPattern(); // returns "Test: {0,number,foo{#.00}";
```

```
var pattIn = "Test: {0,number,foo'}'#.00}";
MessageFormat mFmt = new MessageFormat(pattIn);
var pattOut  = mFmt.toPattern(); // returns "Test: {0,number,foo'}'#.00}";
```

As it stands, it would be inconsistent to have the closing bracket quoted and the opening bracket not quoted. 

Also in response to your earlier question on core-libs-dev, ideally invoking toPattern() can roundtrip, but there are known issues, such as a custom user defined Format subclass, or one of the newer Format subclasses that do not implement the toPattern() method. I am working on making this apparent in the specification of the method in a separate issue.

archiecobbs commented:
Hi @justin-curtis-lu,

Thanks a lot for taking a look, I am glad for any other set of eyes on this tricky stuff!

> As it stands, it would be inconsistent to have the closing bracket quoted and the opening bracket not quoted.

You're right - the problem exists with both `{` and `}`, as is shown here (unmodified jshell):

```
$ jshell
|  Welcome to JShell -- Version 17.0.9
|  For an introduction type: /help intro

jshell> import java.text.*;

jshell> new MessageFormat("Test: {0,number,foo'{'#.00}");
$2 ==> java.text.MessageFormat@951c5b58

jshell> $2.toPattern()
$3 ==> "Test: {0,number,foo{#.00}"

jshell> new MessageFormat($3)
|  Exception java.lang.IllegalArgumentException: Unmatched braces in the pattern.
|        at MessageFormat.applyPattern (MessageFormat.java:521)
|        at MessageFormat.<init> (MessageFormat.java:371)
|        at (#4:1)
```

I've been missing the forest for the trees a bit and I think the fix can be simpler now.

For the record, here is my latest understanding of what's going on...

1. When `MessageFormat.toPattern()` constructs the combined pattern string, it concatenates the plain text bits, suitably escaped, and the pattern strings from each subformat.
1. The subformat strings are already escaped, in the sense that you can take (for example) a `ChoiceFormat` format string and use it as-is to recreate that same `ChoiceFormat`, but they are escaped _only for their own purposes_.
1. The original example is an example of where this matters - `ChoiceFormat` needs to escape `#`, `|`, etc., but doesn't need to escape `{` or `}` - but a `MessageFormat` format string does need to escape those.
1. Therefore, the correct fix is for `MessageFormat.toPattern()` to modify the subformat strings by quoting any _unquoted_ `{` or `}` characters while leaving any already-quoted characters alone.

Updated in a9d78c76f5f. I've also updated the test so it does more thorough round-trip checks.

justin-curtis-lu commented:
Hi @archiecobbs , I think the recent commit is good progress.

First off I think this change will need a CSR as there are behavioral concerns, although minimal. Please let me know if you have access on JBS to file one, otherwise I can file one for you.

> Therefore, the correct fix is for MessageFormat.toPattern() to modify the subformat strings by quoting any unquoted {or } characters while leaving any already-quoted characters alone.

Yes, I think this behavior allows for the String returned by `toPattern()` to create a MessageFormat that can format equivalently to the original MessageFormat. Although to clarify, the original String pattern will not be guaranteed to be equivalent to the one returned by `toPattern()` as we are adding quotes to all brackets in the subformatPattern, regardless if they were quoted in the original String. I think the former is more important here, so the latter is okay.

For DecimalFormat and SimpleDateFormat subformats, adding quotes to brackets found in the subformatPattern is always right, those subformats can not be used to create recursive format behavior. Thus you would **never** except a nested ArgumentIndex in the subformatPattern (ex: `"{0,number,{1}foo0.0"}`), and can confidently escape all brackets found for these subFormats (which you do).

To me, ChoiceFormat is where there is some concern. Consider the following,

`new MessageFormat("{0,choice,0# {1} {2} {3} }”)`

With this change, invoking `toPattern()` on the created MessageFormat would return 

`"{0,choice,0.0# '{'1'}' '{'2'}' '{'3'}' }"`.

This would technically be incorrect. One would think instead of allowing 3 nested elements, we are now printing the literal `{1} {2} {3}` since the brackets are escaped. But, this is not actually the case. Escaping brackets with a choice subFormat does not function properly. This is due to the fact that a recursive messageFormat is created, but the pattern passed has already lost the quotes.

This means that `new MessageFormat("{0,choice,0.0# '{'1'}' '{'2'}' '{'3'}' }")`.

is still equivalent to `new MessageFormat("{0,choice,0# {1} {2} {3} }”).`

So the behavior of quoting all brackets would still guarantee _the String returned by `toPattern()` to create a MessageFormat that can format equivalently to the original MessageFormat_ but only because the current behavior of formatting with a choice subFormat is technically wrong. I think this is okay, since this incorrect behavior is long-standing, but a CSR would be good to address this.

Please let me know if these points made sense, I’m happy to discuss further.

archiecobbs commented:
Hi @justin-curtis-lu,

Thanks for your comments.

> First off I think this change will need a CSR as there are behavioral concerns, although minimal. Please let me know if you have access on JBS to file one, otherwise I can file one for you.

No problem, I'll add one.

> To me, ChoiceFormat is where there is some concern. Consider the following,

OK I think I understand what you're saying.

The first goal here is to ensure the round trip operation `MessageFormat` → Pattern string from `toPattern()` → `MessageFormat` takes you back to where you started.

The patch achieves this so we're good so far. I think we agree on this.

> This would technically be incorrect.

I'm not quite understanding what you meant there (and therefore slightly worried)...

In my mind, a key observation here is that the following two things are not equal and therefore shouldn't be conflated:
* A `ChoiceFormat` pattern string (e.g., what you get from invoking `ChoiceFormat.toPattern()`)
* The `XXX` that you see inside a `MessageFormat` pattern string as part of a `{0,choice,XXX}` subformat.

Those are not the same, even modulo quoting. As you point out, `MessageFormat` has this quirky logic ([here](https://github.com/openjdk/jdk/blob/81df265e41d393cdde87729e091dd465934071fd/src/java.base/share/classes/java/text/MessageFormat.java#L1327-L1333)) in which, after it evaluates a `ChoiceFormat` subformat, if the returned string contains a `{` character, then that string is magically reinterpreted as a `MessageFormat` pattern string and evaluated again, otherwise it's left alone.

So the `XXX` is not a "`ChoiceFormat` pattern string" but something else, so it can't be "incorrect" for not looking like a normal `ChoiceFormat` pattern string.

Maybe it's just a manner of speaking? In any case I want to be careful I'm not missing something in what you're saying.

Thanks.

archiecobbs commented:
/csr

justin-curtis-lu commented:
Hi Archie,

> The XXX that you see inside a MessageFormat pattern string as part of a {0,choice,XXX} subformat.

I think we both agree that XXX may not be the same as the Subformat pattern string.

Sorry if I'm going over stuff you already know, but to clarify my previous statement, MessageFormat escapes brackets so that they are syntactically not evaluated and simply injected into the pattern as a literal bracket. Why I brought up ChoiceFormat, is that it is unique in the sense that you can insert a `{ FormatElement }` into a subFormat pattern, which causes recursion.

Take the example from the class specification, 

`new MessageFormat(There {0,choice,0#are no files|1#is one file|1<are {0,number,integer} files}.”);`

Formatting the value `5`, MessageFormat will first evaluate it as `are {0,number,integer} files}` and then recursively create another MessageFormat to evaluate it as `5 files`.

This is because `1<are {0,number,integer} files.` has unquoted brackets, signaling that MessageFormat should interpret them as syntactically significant. Had they been quoted such as, `1<are '{'0,number,integer'}' files.}`, **I** would expect that MessageFormat will first evaluate it as `are ‘{'0,number,integer’}' files.`, and as the brackets are quoted, interpret them as syntactically meaningless and simply return the value `are {0,number,integer} files.` with no recursive MessageFormat created.


So bringing it back to my example in the previous comment, if we had `new MessageFormat("{0,choice,0.0#foo {1} {2} {3} }")` I we would expect it to format `5 6 7 8` as `foo 6 7 8`. 

But if we had `new MessageFormat("{0,choice,0.0#foo '{'1'}' '{'2'}' '{'3'}' }")` which implies the brackets are syntactically meaningless, I would expect the class to treat them not as nested `FormatElement`s and expect the returned value `foo {0} {1} {2}`

But this is not the case, since by the time the recursive MessageFormat is created the quotes are lost, so it has no means of knowing to treat them as syntactically insignificant, and thus parses the quote and treats them as injected `FormatElement`s which is why in reality the most recent example returns the value `foo 6 7 8`.

So with this change, although calling `toPattern()` on `new MessageFormat("{0,choice,0.0#foo {1} {2} {3} }")` would return the String `"{0,choice,0.0#foo '{'1'}' '{'2'}' '{'3'}' }"` which **I** would think, formats differently, **in reality**, they format equivalently, so I think it is okay. I brought it up not necessarily as an issue, but just something to make apparent. Although I suppose something like this is up for interpretation, so perhaps its only wrong to me.

I'm taking a look at the CSR you filed, but you may also want someone with Reviewer status to take a look as well.
I'll also test your changes against our internal tests. Hope this helped.

archiecobbs commented:
> Sorry if I'm going over stuff you already know...

No apology needed, this stuff is obscure detail madness!

> So with this change, although calling `toPattern()` on `new MessageFormat("{0,choice,0.0#foo {1} {2} {3} }")` would return the String `"{0,choice,0.0#foo '{'1'}' '{'2'}' '{'3'}' }"` which **I** would think, formats differently, **in reality**, they format equivalently, so I think it is okay. I brought it up not necessarily as an issue, but just something to make apparent. Although I suppose something like this is up for interpretation, so perhaps its only wrong to me.

OK gotcha. I totally agree that what you think it should do makes a lot more sense that what it actually does!

archiecobbs commented:
Adding an additional note of explanation here just for the record (this is copied from the CSR):

> We need to ask how do we know it is safe to quote the unquoted curly brace characters in the subformat patterns? If curly braces are not special to the subformat, then quoting them clearly does no harm. So we only need worry about subformat patterns where curly braces are special. But the only subformat pattern strings supported by `MessageFormat` are for `DecimalFormat`, `SimpleDateFormat`, and `ChoiceFormat`, and curly braces are not special for any of these classes, so we're good.
> 
> However, it should be noted that there is some confusing special logic that clouds this question. If the string that results from evaluating a `ChoiceFormat` subformat of a `MessageFormat` contains an opening curly brace, then a new `MessageFormat` is created from that string and evaluated, and _that_ string replaces the original. This behavior doesn't impact how subformats should be quoted, only how their results are interpreted at "run time".

justin-curtis-lu commented on src/java.base/share/classes/java/text/MessageFormat.java:
For this file, please also bump the latter copyright year to 2024.

archiecobbs replied:
Will do.

justin-curtis-lu commented on src/java.base/share/classes/java/text/MessageFormat.java:
```diff
@@ -553,6 +553,9 @@ private void applyPatternImpl(String pattern) {
      * The string is constructed from internal information and therefore
      * does not necessarily equal the previously applied pattern.
      *
+     * @implNote The string returned by this method can be used to create
```
Hmm, I'm not sure about the current note, because its not true in all cases (for example, some unknown subclass of Format). Maybe @naotoj has thoughts?

naotoj replied:
Yes, my comment to the CSR was to explain the default implementation as the `@implNote`. Making it explicit would be helpful to readers.

archiecobbs replied:
Yes my mistake, I'll change this to:
```
@implNote The implementation in {@link MessageFormat} returns a string
that can be used to create a new instance that is semantically equivalent
to this instance.
```

justin-curtis-lu commented on src/java.base/share/classes/java/text/MessageFormat.java:
```diff
@@ -577,9 +581,11 @@ public String toPattern() {
                     if (fmt instanceof DecimalFormat) {
-                        result.append(",number,").append(((DecimalFormat)fmt).toPattern());
+                        result.append(",number");
+                        subformatPattern = ((DecimalFormat)fmt).toPattern();
```
Could use pattern matching `instanceof` here and in the other instances, but I understand if you're trying to stay consistent with the existing code.

archiecobbs replied:
Agreed! Old habits :) I'll fix.

justin-curtis-lu commented on test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java:
```diff
@@ -0,0 +1,323 @@
+public class MessageFormatToPatternTest {
+
+    private static final int NUM_RANDOM_TEST_CASES = 1000;
+    private static final int MAX_FORMAT_NESTING = 3;
```
Can you add a comment defining `MAX_FORMAT_TESTING`, might not be immediately understood without further investigation

archiecobbs replied:
Will do.

justin-curtis-lu commented on test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java:
```diff
@@ -0,0 +1,323 @@
+
+        // A few more test cases from the PR#17416 discussion
+        testRoundTrip(new MessageFormat("Test: {0,number,foo'{'#.00}"));
+        testRoundTrip(new MessageFormat("Test: {0,number,foo'}'#.00}"));
```
Can we add some more concrete test cases, for example, escaped quotes within a MessageFormat pattern (represented by doubled single quotes). 

Another funny case is something like a MessageFormat created by `"{0,number,' abc }'' ' 0.00}"` which becomes `"{0,number, abc '}'''  #0.00}"` after created from the value returned by toPattern() from the first MessageFormat. The Strings appear really different but format equivalently, it would be nice to have suspicious cases like these explicitly defined.

archiecobbs replied:
Sure thing - added in 58e8cc68f45.

justin-curtis-lu commented on test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java:
```diff
@@ -0,0 +1,323 @@
+    // Go roundrip from MessageFormat -> pattern string -> MessageFormat and verify equivalence
+    @ParameterizedTest
+    @MethodSource("testCases")
+    public void testRoundTrip(MessageFormat format1) {
```
Can we also include the original String pattern that created format1, to help with debugging. I find myself wondering what the original String was.

Since technically, the full round trip is _pattern string -> MessageFormat1 -> pattern string -> MessageFormat2_

archiecobbs replied:
I would have done that but it's not (easily) possible. The `MessageFormat`'s are created not from format strings, but by piecing together plain text and sub-`Format` objects manually. This was we are sure what we're dealing with.

Trying to create format strings with multiple levels of nesting from scratch is too complex for my brain due to all the levels of quoting required.

justin-curtis-lu replied:
Right, should have noted that, definitely not worth to try and re-synthesize the original String pattern for each Format. If we're adding some additional concrete cases, its fine since those will clearly have the original String patterns there.

justin-curtis-lu commented on test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java:
```diff
@@ -0,0 +1,323 @@
+
+    // Generate a "random" MessageFormat. We do this by creating a MessageFormat with "{0}" placeholders
+    // and then substituting in random DecimalFormat, DateFormat, and ChoiceFormat subformats. The goal here
+    // is to avoid using pattern strings to construct formats, because they're what we're trying to check.
```
Can you add a general example of a randomly generated MessageFormat in the comments, (I know they vary by quite a lot), but it would be hard for someone to piece if together without digging into the code. And unfortunately the current toString() of MessageFormat is not implemented, so although JUnit prints the MessageFormat out, it's just gibberish.

archiecobbs replied:
Added in 58e8cc68f45 as a test case.

justin-curtis-lu reviewed (commented):
Hi @archiecobbs, left some initial comments.

But I think the logic is sound so far. Quote all quotable characters that need to be quoted, at the MessageFormat level this is simply opening and closing bracket, which you accomplished with the 2 pass copyAndQuoteBraces method. The goal is to create a resultant String that can create a MessageFormat that formats equivalently as the original, (even if calling toPattern() on both MessageFormats returns different String values).

Let's also try and get the CSR approved first.

justin-curtis-lu commented:
Hi Archie, thanks for the updates,

I did testing with some more edge cases, and everything still looks good to me. 

In terms of the CSR,

- Can you update the CSR with the `implSpec` tag as well? 
- Although definitely relevant, I'm not sure if the _special behavior_ section is too involved for a CSR. Perhaps you can move that section to a comment or make it apart of the description in the PR instead?
> After this change, the result of fmt2.toPattern() will be the same as fmt1.toPattern(). 
- While true, I think you should make it apparent that after the change, fmt1.toPattern() returns `{0,number,:'}' #0.##}` now, instead of `{0,number,:} #0.##}`.

Once you make these changes, I will mark it as reviewed by me.

archiecobbs commented:
> In terms of the CSR,...

Hi Justin, thanks again very much for the comments and careful review. The CSR should be updated accordingly.

justin-curtis-lu commented:
Hi Archie,

I wasn't sure if you were waiting to make additional changes or something else, but you can go ahead and re-finalize the CSR (when you are ready), now that it has been reviewed.

archiecobbs commented:
> I wasn't sure if you were waiting to make additional changes or something else, but you can go ahead and re-finalize the CSR (when you are ready), now that it has been reviewed.

Thanks - I wasn't sure about that. I've updated it now.

RogerRiggs commented on src/java.base/share/classes/java/text/MessageFormat.java:
```diff
@@ -553,6 +553,10 @@ private void applyPatternImpl(String pattern) {
      * The string is constructed from internal information and therefore
      * does not necessarily equal the previously applied pattern.
      *
+     * @implSpec The implementation in {@link MessageFormat} returns a string
+     * that can be used to create a new instance that is semantically equivalent
+     * to this instance.
```
The "can be used to create" seems conditional. Perhaps it can be worded as an assertion that can be tested.
```suggestion
     * @implSpec A new MessageFormat created from the string returned from the implementation of 
     * `MessageFormat.toPattern()` is semantically equivalent to this instance.
```

archiecobbs replied:
> The "can be used to create" seems conditional. 

It is conditional - in the sense that you don't _have_ to use it to create a new instance of `MessageFormat`. You can also use it for something else, in other words.

But I also understand how it comes across as a bit wishy-washy...

Hmm, what do you think about this wording?

```
@implSpec The implementation in {@link MessageFormat} returns a string that,
when passed to the {@link MessageFormat(String)} constructor, produces an
instance that is semantically equivalent to this instance.
```

justin-curtis-lu replied:
Not sure which wording will ultimately be used, but if the wording ends up including the constructor, it's probably worth mentioning the `applyPattern` method as well.

archiecobbs replied:
Good point... maybe this?
```
@implSpec The implementation in {@link MessageFormat} returns a string that, when passed
to a {@code MessageFormat()} constructor or {@link #applyPattern applyPattern()}, produces
an instance that is semantically equivalent to this instance.
```

justin-curtis-lu commented on test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java:
```diff
@@ -0,0 +1,360 @@
+
+    @BeforeAll
+    public static void setup() {
+        savedLocale = Locale.getDefault();
```
I'm not sure we need to save the default locale and restore it, unless I'm missing something.

archiecobbs replied:
We are verifying output that includes floating point numbers, and the current locale affects that:
```
jshell> Locale.setDefault(Locale.US);

jshell> new MessageFormat("{0}").format(new Object[] { 1.23 });
$9 ==> "1.23"

jshell> Locale.setDefault(Locale.FRENCH);

jshell> new MessageFormat("{0}").format(new Object[] { 1.23 });
$11 ==> "1,23"
```

justin-curtis-lu replied:
Right, that makes sense.

justin-curtis-lu commented on test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java:
```diff
@@ -0,0 +1,360 @@
+
+/*
+ * @test
+ * @summary Verify MessageFormat.toPattern() is equivalent to original pattern
```
Could make the summary a little more specific for future readers

archiecobbs replied:
Thanks, should be fixed.

justin-curtis-lu commented on test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java:
```diff
@@ -0,0 +1,360 @@
+            // This is the test case from JDK-8323699
+            Arguments.of("{0,choice,0.0#option A: {0}|1.0#option B: {0}'}'}", "option B: 1.23}"),
+            Arguments.of("{0,choice,0.0#option A: {0}|2.0#option B: {0}'}'}", "option A: 1.23"),
+
```
```suggestion
// Absurd double quote examples
Arguments.of("Foo '}''''''''}' {0,number,bar'}' '}' } baz ", "Foo }''''} bar} } 1 baz "),
Arguments.of("'''}''{'''}''''}'"), "'}'{'}''}"),

```

archiecobbs replied:
Thanks, should be fixed.

justin-curtis-lu replied:
Thanks for correcting the suggested test case, second argument had the extra ')' on accident.

justin-curtis-lu reviewed (approved):
Left some minor comments, but LGTM, I also ran this change against internal tests and all is good there. Others may have suggestions, in any case, you will still need a review from someone with 'reviewer' status.

archiecobbs commented on src/java.base/share/classes/java/text/MessageFormat.java:
```diff
@@ -1640,6 +1657,51 @@ private static void copyAndFixQuotes(String source, int start, int end,
+                    qchars.add(new Qchar('\'', quoted));
+                    i++;
+                } else
+                    quoted = !quoted;
```
Thanks, should be fixed.

archiecobbs commented on test/jdk/java/text/Format/MessageFormat/MessageRegression.java:
```diff
@@ -114,9 +114,9 @@ public void Test4074764() {
     @Test
```
Thanks, should be fixed.

archiecobbs commented on test/jdk/java/text/Format/MessageFormat/MessageFormatsByArgumentIndex.java:
```diff
@@ -35,6 +35,7 @@
 public class MessageFormatsByArgumentIndex {
```
Thanks, should be fixed.

naotoj reviewed (approved):
LGTM

justin-curtis-lu reviewed (approved):
Thanks for making the changes

archiecobbs commented:
/integrate

justin-curtis-lu commented:
/sponsor
